hpx/runtime/threads/thread_data.hpp

```diff
@@ -517,7 +516,7 @@ namespace hpx { namespace threads
         coroutine_type::result_type operator()()
         {
             HPX_ASSERT(this == coroutine_.get_thread_id().get());
-            return coroutine_(set_state_ex(wait_signaled));
+            return coroutine_();
```

Comment:
Shouldn't `set_state_ex(wait_signaled)` still be called here (before the call to the coroutine)?

Reply:
There are currently 3 state transitions:
 - (initialized) -> wait_signaled (not a state transition): a new thread is always put into the wait_signaled state
 - wait_signaled -> wait_abort (to cancel a thread): No need to set it back to wait_signaled since the task will terminate
 - wait_timeout -> wait_signaled (when a thread was woken up after a timeout): This is correctly handled here: https://github.com/STEllAR-GROUP/hpx/pull/3126/files#diff-1495f49124e718a7579611b3ba828883L325
 
I don't see a reason to reset the state_ex at each coroutine invocation.

Reply:
Ok, makes sense. Thanks for the explanations.